mlir/test/lib/Dialect/ArmSME/TestLowerToArmSME.cpp

```diff
@@ -0,0 +1,99 @@
+//===- TestLowerToArmSME.cpp - Test lowering to ArmSME as a sink pass -----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass for testing the lowering to ArmSME as a
+// generally usable sink pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/ArithToArmSME/ArithToArmSME.h"
+#include "mlir/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.h"
+#include "mlir/Conversion/ArmSMEToSCF/ArmSMEToSCF.h"
+#include "mlir/Conversion/VectorToArmSME/VectorToArmSME.h"
+#include "mlir/Conversion/VectorToSCF/VectorToSCF.h"
+#include "mlir/Dialect/ArmSME/Transforms/Passes.h"
+#include "mlir/Dialect/ArmSVE/Transforms/Passes.h"
+#include "mlir/IR/DialectRegistry.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Pass/PassOptions.h"
+#include "mlir/Transforms/Passes.h"
+
+using namespace mlir;
+
+namespace {
+struct TestLowerToArmSMEOptions
+    : public PassPipelineOptions<TestLowerToArmSMEOptions> {
+  PassOptions::Option<bool> fuseOuterProducts{
+      *this, "fuse-outer-products",
+      llvm::cl::desc("Fuse outer product operations via "
+                     "'-arm-sme-outer-product-fusion' pass"),
+      llvm::cl::init(true)};
+};
+
```

Comment:
Could you leave a comment with the full pipeline in a form that can be passed `mlir-opt`? I'm pretty sure in the future we'll want to try different orders/passes, so I think that'd be helpful.

Reply:
`-dump-pass-pipeline` can be used for that? This is what it looks like

```
mlir-opt input.mlir -test-lower-to-arm-sme -dump-pass-pipeline

Pass Manager with 13 passes:
builtin.module(arm-sme-vector-legalization,canonicalize{  max-iterations=10 max-num-rewrites=-1 region-simplify=true test-convergence=false top-down=true},cse,convert-arith-to-arm-sme,convert-vector-to-arm-sme,func.func(arm-sme-outer-product-fusion),convert-arm-sme-to-scf,convert-vector-to-scf{full-unroll=true lower-tensors=false target-rank=1},func.func(allocate-arm-sme-tiles),func.func(enable-arm-streaming{only-if-required-by-ops=true streaming-mode=streaming-locally za-mode=new-za}),convert-arm-sme-to-llvm,canonicalize{  max-iterations=10 max-num-rewrites=-1 region-simplify=true test-convergence=false top-down=true},cse)
```

Reply:
I think that's still more noisy than I'd like, I'd prefer the simple form (e.g. `-convert-vector-to-arm-sme -convert-arith-to-arm-sme ...`).

Reply:
copy/pasting:
```
  // -arm-sme-vector-legalization -canonicalize -cse -convert-arith-to-arm-sme
  // -convert-vector-to-arm-sme -arm-sme-outer-product-fusion
  // -convert-arm-sme-to-scf -convert-vector-to-scf=full-unroll
  // -allocate-arm-sme-tiles
  // -enable-arm-streaming="streaming-mode=streaming-locally za-mode=new-za
  // only-if-required-by-ops" -convert-arm-sme-to-llvm -canonicalize -cse
```
seems less convenient than `-dump-pass-pipeline` to me. There's a bit of noise mostly from canonicalization options but it's not the end of the world.

Reply:
Feel free to land this and we can see how things go :slightly_smiling_face:

Reply:
> I think that's still more noisy than I'd like, I'd prefer the simple form (e.g. -convert-vector-to-arm-sme -convert-arith-to-arm-sme ...).

This seems hardly like a reason to encode this in the codebase IMO, this isn't really maintainable.

More importantly the textual pipeline is the only real correct form to describe a pipeline: the short one is only a convenience for running single passes but ambiguous when forming a more complex pipeline.
The option parsing for example can be very confusing when the same pass is attempted to be executed multiple time with different options.

If you would like to skip the default value for the options that aren't touched, this is likely doable: patch welcome for -dump-pass-pipeline here!